xarray/test/test_backends.py

```diff
@@ -1063,3 +1063,131 @@ def test_extract_h5nc_encoding(self):
                           {'least_sigificant_digit': 2})
         with self.assertRaisesRegexp(ValueError, 'unexpected encoding'):
             _extract_nc4_encoding(var, raise_on_invalid=True)
+
+
+class MiscObject:
+    pass
+
+
+class TestValidateAttrs(TestCase):
+
+    def test_invalid_attr_names(self):
+        ds = Dataset({'data': ('y', np.arange(10.0))})
+        self.assert_invalid_attr_names_for_obj(ds.attrs, ds)
+
+        ds = Dataset({'data': ('y', np.arange(10.0))})
+        self.assert_invalid_attr_names_for_obj(ds.data.attrs, ds)
+
+        ds = Dataset({'data': ('y', np.arange(10.0))})
+        self.assert_invalid_attr_names_for_obj(ds.coords['y'].attrs, ds)
+
+    def test_invalid_attr_values(self):
+        ds = Dataset({'data': ('y', np.arange(10.0))})
+        self.assert_invalid_attr_values_for_obj(ds.attrs, ds)
+
+        ds = Dataset({'data': ('y', np.arange(10.0))})
+        self.assert_invalid_attr_values_for_obj(ds.data.attrs, ds)
+
+        ds = Dataset({'data': ('y', np.arange(10.0))})
+        self.assert_invalid_attr_values_for_obj(ds.coords['y'].attrs, ds)
+
+    def test_valid_attr_values(self):
+        ds = Dataset({'data': ('y', np.arange(10.0))})
+        self.assert_valid_attr_values_for_obj(ds.attrs, ds)
+
+        ds = Dataset({'data': ('y', np.arange(10.0))})
+        self.assert_valid_attr_values_for_obj(ds.data.attrs, ds)
+
+        ds = Dataset({'data': ('y', np.arange(10.0))})
+        self.assert_valid_attr_values_for_obj(ds.coords['y'].attrs, ds)
+
+    def assert_invalid_attr_names_for_obj(self, obj, ds):
+        # Helper function for `test_invalid_attr_names`
+        # In this function `obj` is an 'attrs' object (eg. ds.attrs or
+        # data.var.attrs, and `ds` is the dataset object
+
+        obj.clear()
+        obj[123] = 'test'
```

Comment:
I would suggest a slightly different design. Rather than using mutation/`obj.clear()`, use functions to setup brand new objects each time, e.g.,
```
def new_dataset():
    return Dataset({'data': ('y', np.arange(10.0))})

def new_dataset_and_dataset_attrs():
    ds = new_dataset()
    return ds, ds.attrs

def new_dataset_and_data_attrs():
    ds = new_dataset()
    return ds, ds.data.attrs
...

for new_dataset_and_attrs in [new_dataset_and_dataset_attrs, ....]:
    ds, attrs = new_dataset_and_attrs()
    attrs[123] = 'test'
    with self.assertRaisesRegexp(...):
        ...
```
This is a little safer, because using new objects means you can't forget to clean up old ones.